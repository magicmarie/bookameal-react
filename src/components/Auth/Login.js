import React from "react";
import { notify } from "react-notify-toast";
import jwtDecode from "jwt-decode";
import { Link } from "react-router-dom";
import { AppContext } from "../../appContext";
import axiosInstance from "../common/Apicalls";

/**
 *@param {Event} event
 * @class Login
 * @extends {React.Component}
 */
export class Login extends React.Component {
  state = {
    email: "",
    password: "",
    errors: {}
  };

  onChange = event =>
    this.setState({ [event.target.name]: event.target.value });

  // login
  handleLogin = event => {
    event.preventDefault();
    const { email, password } = this.state;
    axiosInstance
      .post("/auth/login", { email, password })
      .then(response => {
        // save token to localstorage, decode to get is_admin value
        localStorage.setItem("token", response.data.token);
        this.props.context.login(response.data.token);
        const decoded = jwtDecode(response.data.token);
        // admin/caterer
        notify.show(response.data.message, "success", 2000);
        const redirectUrl = decoded.is_admin === "True" ? "/admin" : "/user";
        this.props.history.push(redirectUrl);
      })
      .catch(error => {
        if (error.response) {
          notify.show(error.response.data.message, "error", 2500);
        }
      });
  };

  /**
   * @returns {any} rendered items
   * @memberof Login
   */
  render() {
    const { email, password } = this.state;
    return (
      <div className="row justify-content-md-center">
        <div className="col-md-5">
          <div className="jumbotron">
            <form onSubmit={this.handleLogin}>
              <h2 className="text-center form-signin-heading">
                Log in to your account
              </h2>
              <div className="form-group">
                <label htmlFor="email">Email</label>
                <input
                  type="email"
                  value={email}
                  onChange={this.onChange}
                  name="email"
                  required
                  className="form-control"
                />
              </div>
              <div className="form-group">
                <label htmlFor="password">Password</label>
                <input
                  type="password"
                  value={password}
                  onChange={this.onChange}
                  name="password"
                  required
                  className="form-control"
                />
              </div>
              <div>
                <button type="submit" className="btn btn-primary btn-block">
                  Login
                </button>
              </div>
              <p />
              <p className="text-center">
                Do not have an account yet?
                <Link href="/signup" to="/signup">
                  {" "}
                  Sign Up
                </Link>
              </p>
            </form>
          </div>
        </div>
      </div>
    );
  }
}

export default React.forwardRef((props, ref) => (
  <AppContext.Consumer>
    {context => <Login {...props} context={context} ref={ref} />}
  </AppContext.Consumer>
));
