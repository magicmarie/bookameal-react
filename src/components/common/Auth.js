// AuthRoute: to protect authenticated user routes: user
import React from "react";
import { Redirect, Route } from "react-router-dom";

const AuthRoute = ({ component: Component, ...rest }) => {
  // check the localStorage for the token: if missing, redirect to login.
  // otherwise load the component
  const token = localStorage.getItem("token");
  return (
    <Route
      {...rest}
      render={props => {
        return token === null ? (
          <Redirect to="/login" />
        ) : (
          <Component {...props} />
        );
      }}
    />
  );
};

export default AuthRoute;
