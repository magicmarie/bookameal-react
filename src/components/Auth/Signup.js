import React from "react";
import { notify } from "react-notify-toast";
import { Link } from "react-router-dom";
import axiosInstance from "../common/Apicalls";
import { Form } from "../common/Helper";

/**
 *@param {Event} event
 * @class Signup
 * @extends {React.Component}
 */
class Signup extends React.Component {
  state = {
    name: "",
    email: "",
    password: "",
    is_admin: false,
    errors: {}
  };

  onChange = event => {
    // if name is is_admin, do this
    if (event.target.name === "is_admin") {
      this.setState({
        is_admin: event.target.checked
      });
    } else {
      this.setState({
        [event.target.name]: event.target.value
      });
    }
  };

  // sign up
  handleSignup = event => {
    event.preventDefault();
    const { name, email, password, is_admin, errors } = this.state;
    axiosInstance
      .post("/auth/signup", { name, email, password, is_admin, errors })
      .then(response => {
        notify.show(response.data.message, "success", 2500);
        this.props.history.push("/login");
      })
      .catch(err => {
        if (err.response) {
          notify.show(err.response.data.message, "error", 2500);
        }
      });
  };

  /**
   * @returns {any} rendered items
   * @memberof Signup
   */
  render() {
    const { email, name, password, is_admin } = this.state;
    const formFields = [
      {
        type: "text",
        name: "name",
        value: name,
        label: "Name"
      },
      {
        type: "email",
        name: "email",
        value: email,
        label: "Email"
      },
      {
        type: "password",
        name: "password",
        value: password,
        label: "Password"
      }
    ];
    return (
      <div className="row justify-content-md-center">
        <div className="col-md-5">
          <div className="jumbotron">
            <form onSubmit={this.handleSignup}>
              <h3 className=" text-center form-signin-heading">
                New user, Create an account
              </h3>
              <Form formFields={formFields} onChange={this.onChange} />
              <div className="form-check form-group">
                <input
                  type="checkbox"
                  name="is_admin"
                  checked={is_admin}
                  onChange={this.onChange}
                  className="form-check-input"
                />
                <label className="form-check-label" htmlFor="isadmin">
                  Caterer
                </label>
                <p className="danger">For caterer only</p>
              </div>
              <div>
                <button type="submit" className="btn btn-primary btn-block">
                  Signup
                </button>
              </div>
              <p />
              <p className="text-center">
                Already have an account?
                <Link href="/login" to="/login">
                  {" "}
                  Login
                </Link>
              </p>
            </form>
          </div>
        </div>
      </div>
    );
  }
}

export default Signup;
