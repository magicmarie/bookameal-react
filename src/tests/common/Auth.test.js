import React from "react";
import { shallow } from "enzyme";
import Auth from "../../components/common/Auth";

describe("AuthRoute component", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Auth />);
  });
  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
