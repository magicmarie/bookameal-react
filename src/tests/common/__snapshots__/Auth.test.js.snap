// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthRoute component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthRoute />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "render": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "render": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
